### ISSUE TITLE
Retrieving posts

### ISSUE THREAD

**Issue by silverpill** [anomalous69/FChannel#4]
When I try to retrieve a `Note` by its `id`, the server returns a collection instead of a requested `Note`:

`https://usagi.reisen/b/X11WVVIW`

```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "actor": {
    "id": "https://usagi.reisen/b",
    "restricted": false
  },
  "orderedItems": [
   ...
  ]
}
```

**Comment by hitstill** [comment]
It appears something is not right with ActorPosts() handler. 
I am new to this project's codebase, but will have a closer look when I have some a spare hour or two. No promises though

**Comment by hitstill** [comment]
Apparently that was a design choice: to pull up an entire thread to which desired note belongs to and then let the browser to do the rest.
Try out for yourself with some of the child notes. For example https://usagi.reisen/b/QES3FBNW you will get https://usagi.reisen/b/X11WVVIW#QES3FBNW where X11WVVIW is the top level id, and the #QES3FBNW part displays you what you wanna see.
If you think about it for a while, it actually makes perfect sense.
Not my place to decide but from my perspective that's correct behavior although "deviates" from the specs a little

**Comment by iwojim0** [comment]
hi
>that's correct behavior

I'm not AP guru and dont know thats right or no:)
but if I right understood silverpill - this is one from isues that not allow federate fchannel with another fedi instances.

**Comment by hitstill** [comment]
> but if I right understood silverpill - this is one from isues that not allow federate fchannel with another fedi instances.

Yeah. I get that. What I am trying to say judging by what I saw in the sources that's more like a deliberate design choice rather than a programming error. User experience above spec conformance. 

Here is commit from 2022 made by the previous maintainer which introduced such behavior https://github.com/anomalous69/FChannel/commit/60752f58ab0eade30e88c1e98ad530a4b151586e 
We could try to revert it if @anomalous69 is ok with that approach.

**Comment by silverpill** [comment]
I agree that current behavior makes sense on the frontend (when you open URL with a web browser).
But when ActivityPub representation is served, it could be implemented differently without degrading user experience.

I am retrieving ActivityPub representation with the following command:

```shell
curl -H 'Accept: application/ld+json; profile="https://www.w3.org/ns/activitystreams"' https://usagi.reisen/b/X11WVVIW 
```

It seems that FChannel already knows how to serve different content depending on the value of `Accept` header. Perhaps the returned data can be altered as well?

**Comment by hitstill** [comment]
You are correct in your assumption regarding Accept headers. There are conditionals in every handler. In particular https://github.com/anomalous69/FChannel/blob/c793a28405a1bb46502905a9efabbeae6597cac4/routes/actor.go#L451 is the one of interest. It calls GetActorPost if the header is present. Which I believe supposed to return a single entry instead of collection to comply the spec. And this behavior is clearly a bug.
Thanks for the suggestion. I think I can take it from here

**Comment by silverpill** [comment]
If collection of posts is needed for federation between FChannel instances, it can be attached to top-level Note using a `context` property. Some ActivityPub implementations do this:

https://codeberg.org/fediverse/fep/src/branch/main/fep/f228/fep-f228.md#collection-of-posts

**Comment by hitstill** [comment]
I think I've got ourselves a fix

```
curl -H 'Accept: application/ld+json; profile="https://www.w3.org/ns/activitystreams"' http://localhost:3000/g/NSM7K6X4
```

```
{
	"type": "Note",
	"id": "http://localhost:3000/g/NSM7K6X4",
	"actor": "http://localhost:3000/g",
	"content": "Nice",
	"published": "2025-02-25T16:08:52.557358Z"
}
```
where NSM7K6X4 is a reply.  Web version is unaffected by the changes and outputs complete collection
Still testing. Patch coming up soon

**Comment by silverpill** [comment]
@hitstill Thank you!

**Comment by hitstill** [comment]
I created a new issue for potential 0.2.1 update https://github.com/anomalous69/FChannel/issues/8 containing among other changes fix for this little problem.
For now that's all I can do. The rest is in the hands of this repo maintainer who is I suppose currently busy with other stuff.

**Comment by silverpill** [comment]
@hitstill usagi.reisen has been updated (v0.2.0-cb01df3), but posts still resolve to collections

Here's a new one: https://usagi.reisen/b/WAKOW2BJ

**Comment by hitstill** [comment]
Actually not exactly. It returns a single note with  all of its replies, but for some reason as a single element collection. It appears there is more fixing to do

**Comment by hitstill** [comment]
Quick update: I wasn't able to figure out why it happens by going through the code. 
Going to launch local instance and investigate at run-time tomorrow. Apparently I've missed something, just can't figure out what  is that exactly
